introspect.go

```diff
@@ -0,0 +1,173 @@
+package swarm
+
+import (
+	"github.com/libp2p/go-libp2p-core/introspect"
+	introspectpb "github.com/libp2p/go-libp2p-core/introspect/pb"
+	"github.com/pkg/errors"
+)
+
+// IntrospectTraffic introspects & returns the overall traffic for this peer
+func (s *Swarm) IntrospectTraffic() (*introspectpb.Traffic, error) {
+	if s.bwc != nil {
+		t := &introspectpb.Traffic{}
+		metrics := s.bwc.GetBandwidthTotals()
+
+		t.TrafficIn = &introspectpb.DataGauge{CumBytes: uint64(metrics.TotalIn), InstBw: uint64(metrics.RateIn)}
+		t.TrafficOut = &introspectpb.DataGauge{CumBytes: uint64(metrics.TotalOut), InstBw: uint64(metrics.RateOut)}
+
+		return t, nil
+	}
+
+	return nil, nil
+}
+
+// IntrospectConnections introspects & returns the swarm connections
+func (swarm *Swarm) IntrospectConnections(query introspect.ConnectionQueryParams) ([]*introspectpb.Connection, error) {
+	swarm.conns.RLock()
+	defer swarm.conns.RUnlock()
+
+	containsId := func(ids []introspect.ConnectionID, id string) bool {
```

Comment:
More efficient to create a `map[introspect.ConnectionID]struct{}`, and make the predicate test membership. This is `O(m*n)` worst case scenario; map-based solution would be `O(m)` at the expense of a little extra memory.